### ISSUE TITLE
Error when running Collision Box model with adaptive and double_precision on ThrustRTC (CUDA atomicAdd(double *, double) missing)

### ISSUE THREAD

**Issue by abulenok** [open-atmos/PySDM#1138]
```
[/content/PySDM/PySDM/particulator.py](https://localhost:8080/#) in run(self, steps)
     46             for key, dynamic in self.dynamics.items():
     47                 with self.timers[key]:
---> 48                     dynamic()
     49             self.n_steps += 1
     50             self._notify_observers()

[/content/PySDM/PySDM/dynamics/collisions/collision.py](https://localhost:8080/#) in __call__(self)
    179                 while self.particulator.attributes.get_working_length() != 0:
    180                     self.particulator.attributes.cell_idx.sort_by_key(self.dt_left)
--> 181                     self.step()
    182                     self.particulator.attributes.cut_working_length(
    183                         self.particulator.adaptive_sdm_end(self.dt_left)

[/content/PySDM/PySDM/dynamics/collisions/collision.py](https://localhost:8080/#) in step(self)
    212             proc_rand = None
    213 
--> 214         self.compute_gamma(
    215             prob=prob, rand=rand, is_first_in_pair=self.is_first_in_pair, out=self.gamma
    216         )

[/content/PySDM/PySDM/dynamics/collisions/collision.py](https://localhost:8080/#) in compute_gamma(self, prob, rand, is_first_in_pair, out)
    260         """
    261         if self.adaptive:
--> 262             self.particulator.backend.scale_prob_for_adaptive_sdm_gamma(
    263                 prob=prob,
    264                 multiplicity=self.particulator.attributes["multiplicity"],

[/content/PySDM/PySDM/backends/impl_thrust_rtc/nice_thrust.py](https://localhost:8080/#) in wrapper(*args, **kwargs)
     10             if debug_print:
     11                 print(func.__name__)
---> 12             result = func(*args, **kwargs)
     13             if wait:
     14                 trtc.Wait()

[/content/PySDM/PySDM/backends/impl_thrust_rtc/methods/collisions_methods.py](https://localhost:8080/#) in scale_prob_for_adaptive_sdm_gamma(self, prob, multiplicity, cell_id, dt_left, dt, dt_range, is_first_in_pair, stats_n_substep, stats_dt_min)
    698             n=len(dt_left), args=(dt_todo, dt_left.data, d_dt_max)
    699         )
--> 700         self.__scale_prob_for_adaptive_sdm_gamma_body_2.launch_n(
    701             n=len(multiplicity) // 2,
    702             args=(

[/usr/local/lib/python3.10/dist-packages/ThrustRTC/Context.py](https://localhost:8080/#) in launch_n(self, n, args)
     85     def launch_n(self, n, args):
     86         arg_list = ObjArray(args)
---> 87         check_i(native.n_for_launch_n(self.m_cptr, n, arg_list.m_cptr))

[/usr/local/lib/python3.10/dist-packages/ThrustRTC/Native.py](https://localhost:8080/#) in check_i(ret_value_i)
     14 def check_i(ret_value_i):
     15         if ret_value_i == -1:
---> 16                 raise SystemError("An internal error happend")
     17         elif ret_value_i == -2:
     18                 raise ValueError("Wrong number of arguments.")

SystemError: An internal error happened
```

**Comment by abulenok** [comment]
experienced when running this notebook https://github.com/open-atmos/PySDM/blob/performance-comparison/examples/PySDM_examples/Srivastava_1982/performance_comparison_Srivastava_Setup.ipynb

**Comment by abulenok** [comment]
Box model with Collision dynamic (coalescence + breakup)
adaptive=True
GPU(double_precision=True)

**Comment by slayoo** [comment]
Reproduced locally. The full error log includes the cause of the error:
```
Errors:
header_of_structs.h(865): error: no instance of overloaded function "atomicAdd" matches the argument list
            argument types are: (double *, double)
```
Here's an explanation: https://forums.developer.nvidia.com/t/why-does-atomicadd-not-work-with-doubles-as-input/56429

**Comment by slayoo** [comment]
BTW, CUDA docs mention a workaround (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomic-functions):
```c++
#if __CUDA_ARCH__ < 600
__device__ double atomicAdd(double* address, double val)
{
    unsigned long long int* address_as_ull =
                              (unsigned long long int*)address;
    unsigned long long int old = *address_as_ull, assumed;

    do {
        assumed = old;
        old = atomicCAS(address_as_ull, assumed,
                        __double_as_longlong(val +
                               __longlong_as_double(assumed)));

    // Note: uses integer comparison to avoid hang in case of NaN (since NaN != NaN)
    } while (assumed != old);

    return __longlong_as_double(old);
}
#endif
```

**Comment by slayoo** [comment]
addressed in #1141 (https://github.com/open-atmos/PySDM/pull/1141/commits/b91d3d72be2187667fcb3acb370285f9052cb4fb)
